### DISCUSSION TITLE
apply min/max limits determined by contrast-stretch for xx% on yy%

### DISCUSSION THREAD

**Discussion by crisandhiscrisis** [ImageMagick/ImageMagick#2383] [Help]
Dear community,

I have started working with the image magick command line tool some weeks ago and so far, I found all answers either in its documentation, in scripts written by others or in the old forum. But my current problem yields no results via the usual search engines, so I finally resorted to asking you directly.

I need to perform a contrast-strech operation on some images. I let each channel be adjusted individually by using the "-channel RGB" option. My problem is, that using "2%x2%", the color balance of black and white is adjusted perfectly, but too much is cut off. Applying just "0.1%x0.1%" saves shadow and highlight detail, but color keeps a cast.

How can I estimate the ratio of adjustment between each channel when the 2% criteria is used, so that I can apply the same ratio with a lesser amount of clipping, e.g. 0.5%?

Best wishes,
Chris and his crisis

**Comment by snibgo** [comment]
What version of IM, on what platform?

> I let each channel be adjusted individually by using the "-channel RGB" option.

Why are you doing this? I think you actually want to adjust all channels equally.

"-channel RGB -contrast-stretch XX%xYY%" will often cause colour-shifts. That is inevitable.

"-contrast-stretch XX%xYY%" is equivalent to some "-level AA%xBB%" for each channel, where AA and BB is calculated from the histograms. Note that there are two values, AA and BB, so a simple ratio between channels won't work.

Sadly, IM can't directly tell us the equivalent "-level AA%xBB%". You can create your own histograms, and use those to find your own levels.

**Reply by crisandhiscrisis** [reply]
"You can create your own histograms, and use those to find your own levels."

That's probably the way to go. Can you point me in the direction how to do it? I want to adjust every channel separately to get rid of color casts. Higher percentages work great in terms of neutralizing color casts, but lead to clipping.

I'm running 6.9.7-4 on Linux.

**Reply by snibgo** [reply]
> You can create your own histograms, ...

For that, I would use my process modules. See [Process modules](http://im.snibgo.com/customim.htm), and the entries for mkhisto and invclut on that page. But this needs IM to be re-built, with my source code (on that page).

> I want to adjust every channel separately to get rid of color casts.

This usually means multiplying each channel by a different number, which is equivalent to "-level 0,L%" where L is a different number for each channel.


On histograms:

In its purest form, a histogram is a Nx1 image, where N is the number of buckets and where channel values in the histogram are counts (or proportions of the total pixels) of pixels that have value in that bucket.

A _cumulative_ histogram has cumulative values. We often normalise this so the final count is 100% of quantum.

If we make a cumulative normalised histogram then the input corresponds to a "-level" number, and the output is the proportion of pixels that are beneath that level.

If we invert that histogram, the output is a "-level" and the input is a proportion of pixels. So we can crop a pair of pixels to find the levels for all channels.

```
magick ^
  toes.png ^
  -process 'mkhisto cumul norm' ^
  -process invclut ^
  ( -clone 0 -crop 1x1+%%[fx:w*0.02]+0 +repage ) ^
  ( -clone 0 -crop 1x1+%%[fx:w*0.98]+0 +repage ) ^
  -delete 0 ^
  +append +repage ^
  txt:
```
We find the levels that correspond to "-contrast-stretch 2,98%".

The text output is:
```
# ImageMagick pixel enumeration: 2,1,4294967295,srgb
0,0: (9.89397e+08,7.09296e+08,7.75553e+08)  #3AF92A472E3A  srgb(23.0362%,16.5146%,18.0573%)
1,0: (4.00628e+09,3.40853e+09,3.28932e+09)  #EECACB29C40E  srgb(93.2785%,79.361%,76.5854%)
```
Look at the percentages at the end. If we have a simple "-level 23.0362,93.2785%", the red channel would be be clipped by 2% at each end. A conservative "-level" would be "-level 16.5146,93.2785%", which would clip 2% of green in the shadows and 2% of red in the highlights, with less clipping of the other channels.

**Comment by snibgo** [comment]
I said:
> "-contrast-stretch XX%xYY%" is equivalent to some "-level AA%xBB%" for each channel, where AA and BB is calculated from the histograms.

Okay, but if we know an image has been transformed to another image by "-level AA%xBB%" for each channel, can we calculate AA and BB from the two images? Yes, we can. The transformations are linear, so we need two pixels in the input image, and the two corresponding pixels in the output image. In each image, the two pixels must be different in all channels, and must not have been clamped. The following Windows BAT script does the work, but doesn't check the conditions.

```
rem Given one image is transformed by "-level", and we have the result image, what are the parameters?

rem We compare corresponding pixels in the two images.
rem The pixels must be different, and must not have been clamped.

rem Suppose we had "-level L1,L2" where L1 and L2 are proportions, [0..1].

rem Then vout = (vin-L1)/ (L2-L1)
rem where vin and vout are [0..1].

rem vout0 = (vin0-L1) / (L2-L1) = vin0/(L2-L1) - L1/(L2-L1)
rem vout1 = (vin1-L1) / (L2-L1) = vin1/(L2-L1) - L1/(L2-L1)

rem Subtract:
rem vout0-vout1 = vin0/(L2-L1) - vin1/(L2-L1)
rem             = (vin0-vin1) / (L2-L1)

rem L2-L1 = (vin0-vin1) / (vout0-vout1)

rem L1 = vin0 - vout0 * (L2-L1)

rem Create a test sample.

%IMG7%magick toes.png -level 2x95%% t.png


for /F "usebackq" %%L in (`%IMG7%magick ^
 -precision 15 ^
  toes.png t.png ^
  ^( -clone 0-1 ^
     -crop 1x1+10+10 +repage ^
     ^( -clone 0 ^
        -format "vin0r=%%[fx:mean.r]\nvin0g=%%[fx:mean.g]\nvin0b=%%[fx:mean.b]\n" -write info: ^
        +delete ^
     ^) ^
     ^( -clone 1 ^
        -format "vout0r=%%[fx:mean.r]\nvout0g=%%[fx:mean.g]\nvout0b=%%[fx:mean.b]\n" -write info: ^
        +delete ^
     ^) ^
     -delete 0--1 ^
  ^) ^
  ^( -clone 0-1 ^
     -crop 1x1+%%[fx:w-10]+%%[fx:h-10] +repage ^
     ^( -clone 0 ^
        -format "vin1r=%%[fx:mean.r]\nvin1g=%%[fx:mean.g]\nvin1b=%%[fx:mean.b]\n" -write info: ^
        +delete ^
     ^) ^
     ^( -clone 1 ^
        -format "vout1r=%%[fx:mean.r]\nvout1g=%%[fx:mean.g]\nvout1b=%%[fx:mean.b]\n" -write info: ^
        +delete ^
     ^) ^
     -delete 0--1 ^
  ^) ^
  null:`) do set %%L

set FMT=^
L2mL1r=%%[fx:(%vin0r%-%vin1r%)/(%vout0r%-%vout1r%)]\n^
L2mL1g=%%[fx:(%vin0g%-%vin1g%)/(%vout0g%-%vout1g%)]\n^
L2mL1b=%%[fx:(%vin0b%-%vin1b%)/(%vout0b%-%vout1b%)]\n

for /F "usebackq" %%L in (`%IMG7%magick identify ^
  -precision 15 ^
  -format "%FMT%" ^
  xc:`) do set %%L

set FMT=^
L1r=%%[fx:(%vin0r%-%vout0r%*%L2mL1r%)]\n^
L1g=%%[fx:(%vin0g%-%vout0g%*%L2mL1g%)]\n^
L1b=%%[fx:(%vin0b%-%vout0b%*%L2mL1b%)]\n

for /F "usebackq" %%L in (`%IMG7%magick identify ^
  -precision 15 ^
  -format "%FMT%" ^
  xc:`) do set %%L

set FMT=^
L2r=%%[fx:(%L2mL1r%+%L1r%)]\n^
L2g=%%[fx:(%L2mL1g%+%L1g%)]\n^
L2b=%%[fx:(%L2mL1b%+%L1b%)]\n

for /F "usebackq" %%L in (`%IMG7%magick identify ^
  -precision 15 ^
  -format "%FMT%" ^
  xc:`) do set %%L
```

Results are:
```
f:\web\im>set L1r=0.0200083445290729
f:\web\im>set L1g=0.0199946759016898
f:\web\im>set L1b=0.0199862159434194

f:\web\im>set L2r=0.94999097149364
f:\web\im>set L2g=0.949999189092516
f:\web\im>set L2b=0.950013498883559
```
As we hoped, the levels for the channels are more or less the same, and more or less correct.
